lib/middle.js

```diff
@@ -0,0 +1,258 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ * @private
+ */
+
+var EventEmitter = require('events').EventEmitter;
+var finalhandler = require('finalhandler');
+var merge = require('utils-merge');
+var parseUrl = require('parseurl');
+
+/**
+ * Module exports.
+ * @public
+ */
+
+module.exports = createServer;
+module.exports.extractHandlers = extractHandlers;
+
+/**
+ * Extract handlers from a middle instance
+ *
+ * @function extractHandlers
+ * @param  {Middle} middle - middle instance
+ * @returns {Function[]} handlers
+ */
+function extractHandlers(middle) {
+    return middle.stack.map(function map(stackItem) {
+        return stackItem.handle;
+    });
+}
+
+/**
+ * Module variables.
+ * @private
+ */
+
+var env = process.env.NODE_ENV || 'development';
+var proto = {};
+
+/* istanbul ignore next */
+var defer =
```

Comment:
Why are we checking for the existence of `setImmediate`?

Reply:
Sry, I didn't make it clear that this is `connect`, I copy-pasted it, because I needed to extend it with the `next(false) logic. :(  Later we can try to open a PR or replace `next(false)` with something else.